### ISSUE TITLE
Adding Coregion Kernel

### ISSUE THREAD

**Issue by tirthasheshpatel** [JuliaGaussianProcesses/KernelFunctions.jl#128]
Coregion kernel is used in coregionalization GP models and AFAIK it is only present in gpy, pymc3, and gpflow. Would it be helpful? If so, I can take a shot at it.

I am not sure how AD would work. [Here's the link to GPy's implementation](http://gpy.readthedocs.io/en/deploy/GPy.kern.src.html#GPy.kern.src.coregionalize.Coregionalize)

**Comment by willtebbutt** [comment]
Hiya! Yeah, this is something that's on the TODO list for multi-output GPs. Please see [this](https://github.com/JuliaGaussianProcesses/AbstractGPs.jl/issues/9) discussion regarding multi-output GPs. Would be great if you could have a stab at it, but I think our approach to it is going to be slightly non-standard.

**Comment by tirthasheshpatel** [comment]
I am new to Julia and facing some problems setting up the environment. Is there a standard guide to build and test this package?

**Comment by tirthasheshpatel** [comment]
<details>

<summary>I have something like this on my mind:</summary>

```julia
using LinearAlgebra

struct CoregionKernel{T<:Real} <: BaseKernel
    B::AbstractMatrix{T}
    CoregionKernel(; B::AbstractMatrix{T}) where {T<:Real} = new{T}(B)
    CoregionKernel(; W::AbstractMatrix{T}, κ::AbstractVector{T}) where {T<:Real} = new{T}(W*W' + diagm(κ))
end

function (κ::CoregionKernel)(x::AbstractVector{<:Real}, y::AbstractVector{<:Real})
    @assert length(x) ≠ 1 "Only one observed dimension allowed"
    idx = convert.(AbstractVector{Int}, floor.(x))
    idy = convert.(AbstractVector{Int}, floor.(y))
    return κ.B[idx, idy]
end

function (κ::CoregionKernel)(x::Real, y::Real)
    idx = convert(Int, floor(x))
    idy = convert(Int, floor(y))
    return κ.B[idx, idy]
end

Base.show(io::IO, κ::CoregionKernel) = print(io, "Coregion Kernel (B = ", κ.B, ")")

function kernelmatrix(κ::CoregionKernel, x::AbstractVector{T}) where {T<:Real}
    @assert size(x) ≠ 1 "Only one observed dimension allowed"
    idx = convert.(AbstractVector{Int}, floor.(x))
    return κ.B[idx, idx']
end

function kernelmatrix!(K::AbstractMatrix{T}, κ::CoregionKernel, x::AbstractVector{T}) where {T<:Real}
    @assert size(x) ≠ 1 "Only one observed dimension allowed"
    idx = convert.(AbstractVector{Int}, floor.(x))
    K .= κ.B[idx, idx']
    return K
end

function kernelmatrix(κ::CoregionKernel, x::AbstractVector{T}, y::AbstractVector{T}) where {T<:Real}
    @assert size(x) ≠ 1 "Only one observed dimension allowed"
    idx = convert.(AbstractVector{Int}, floor.(x))
    idy = convert.(AbstractVector{Int}, floor.(y))
    return κ.B[idx, idy']
end

function kernelmatrix!(
    K::AbstractMatrix,
    κ::CoregionKernel,
    x::AbstractVector,
    y::AbstractVector,
)
    @assert size(x) ≠ 1 "Only one observed dimension allowed"
    idx = convert.(AbstractVector{Int}, floor.(x))
    idy = convert.(AbstractVector{Int}, floor.(y))
    K .= κ.B[idx, idy']
    return K
end
```

</details>

If I understood the current implementation correctly, the above snippet should do. What approach are you thinking of?

**Comment by theogf** [comment]
To install a package go in the REPL and type `]add KernelFunctions` for example. 
If you want to develop it you can fork it on github and type `]dev https://yourgithubfork`

**Comment by devmotion** [comment]
Maybe the documentation of [Pkg](https://julialang.github.io/Pkg.jl/v1/managing-packages/) might be helpful for setting up and testing packages in general.

**Comment by tirthasheshpatel** [comment]
Thanks @devmotion and @theogf. I will try it out and get back to you.

**Comment by theogf** [comment]
> I have something like this on my mind:
> ```julia
> using LinearAlgebra
> 
> struct CoregionKernel{T<:Real} <: BaseKernel
>     B::AbstractMatrix{T}
>     CoregionKernel(; B::AbstractMatrix{T}) where {T<:Real} = new{T}(B)
>     CoregionKernel(; W::AbstractMatrix{T}, κ::AbstractVector{T}) where {T<:Real} = new{T}(W*W' + diagm(κ))
> end
> 
> function (κ::CoregionKernel)(x::AbstractVector{<:Real}, y::AbstractVector{<:Real})
>     @assert length(x) ≠ 1 "Only one observed dimension allowed"
>     idx = convert.(AbstractVector{Int}, floor.(x))
>     idy = convert.(AbstractVector{Int}, floor.(y))
>     return κ.B[idx, idy]
> end
> 
> function (κ::CoregionKernel)(x::Real, y::Real)
>     idx = convert(Int, floor(x))
>     idy = convert(Int, floor(y))
>     return κ.B[idx, idy]
> end
> 
> Base.show(io::IO, κ::CoregionKernel) = print(io, "Coregion Kernel (B = ", κ.B, ")")
> 
> function kernelmatrix(κ::CoregionKernel, x::AbstractVector{T}) where {T<:Real}
>     @assert size(x) ≠ 1 "Only one observed dimension allowed"
>     idx = convert.(AbstractVector{Int}, floor.(x))
>     return κ.B[idx, idx']
> end
> 
> function kernelmatrix!(K::AbstractMatrix{T}, κ::CoregionKernel, x::AbstractVector{T}) where {T<:Real}
>     @assert size(x) ≠ 1 "Only one observed dimension allowed"
>     idx = convert.(AbstractVector{Int}, floor.(x))
>     K .= κ.B[idx, idx']
>     return K
> end
> 
> function kernelmatrix(κ::CoregionKernel, x::AbstractVector{T}, y::AbstractVector{T}) where {T<:Real}
>     @assert size(x) ≠ 1 "Only one observed dimension allowed"
>     idx = convert.(AbstractVector{Int}, floor.(x))
>     idy = convert.(AbstractVector{Int}, floor.(y))
>     return κ.B[idx, idy']
> end
> 
> function kernelmatrix!(
>     K::AbstractMatrix,
>     κ::CoregionKernel,
>     x::AbstractVector,
>     y::AbstractVector,
> )
>     @assert size(x) ≠ 1 "Only one observed dimension allowed"
>     idx = convert.(AbstractVector{Int}, floor.(x))
>     idy = convert.(AbstractVector{Int}, floor.(y))
>     K .= κ.B[idx, idy']
>     return K
> end
> ```
> 
> If I understood the current implementation correctly, the above snippet should do. What approach are you thinking of?

Nice effort! Unfortunately I think it would be a bit more complex than that. A `CoregionKernel` would first need to contain a kernel in the first place.
Then the API should more straightforward on what should the kernel receive, inputs, what output etc, again we need to agree on this before starting implementing it. It would be better to continue this discussion in https://github.com/JuliaGaussianProcesses/AbstractGPs.jl/issues/9

**Comment by 4aHxKzD** [comment]
I would like to contribute by adding this kernel if possible, but I'm still not 100% comfortable with the API you guys designed. What do you guys think should be improved and what is missing from this first draft?

```julia
struct IntrinsicCoregionalizationMOKernel{Tu<:Kernel, TA<:AbstractMatrix, Tv<:LinearAlgebra.Diagonal, TB<:AbstractMatrix, Trank<:Integer, Tnoutputs<:Integer} <: MOKernel
    u::Tu
    A::TA
    v::Tv
    B::TB
    rank::Trank
    noutputs::Tnoutputs

    function IntrinsicCoregionalizationMOKernel(u, noutputs, rank)
        noutputs >= rank || error("`noutputs` should be greater or equal than `rank`")

        A = rand(noutputs, rank)
        v = Diagonal(rand(noutputs))
        B = A * transpose(A) + v

        return new{typeof(u), typeof(A),  typeof(v), typeof(B), typeof(rank), typeof(noutputs)}(u,A,v,B,rank,noutputs)
    end

    function IntrinsicCoregionalizationMOKernel(u, A, v::AbstractVector)
        rank = size(A,2)
        noutputs = size(A,1)
        v = Diagonal(v)
        B = A * transpose(A) + v
        return new{typeof(u), typeof(A),  typeof(v), typeof(B), typeof(rank), typeof(noutputs)}(u,A,v,B,rank,noutputs)
    end
end



function (κ::IntrinsicCoregionalizationMOKernel)((x, px)::Tuple{Any,Int}, (y, py)::Tuple{Any,Int})
    return κ.B[px,py] * κ.u(x,y)
end


# function kernelmatrix(κ::IntrinsicCoregionalizationMOKernel,
#                       x::AbstractVector{Tuple{Any,Int}},
#                       y::AbstractVector{Tuple{Any,Int}})

#     return κ.(x, permutedims(y))
# end


function Base.show(io::IO, κ::IntrinsicCoregionalizationMOKernel)
    return print(io, "Intrinsic Coregionalization Multi-Output Kernel")
end
```

**Comment by 4aHxKzD** [comment]
Should I open a pull request from https://github.com/4aHxKzD/KernelFunctions.jl instead of discussing in this issue?

**Comment by devmotion** [comment]
> What do you think should be improved and what is missing from this first draft?

I think the basic kernel would only work with `B`. I.e., 
```julia
struct CoregionMOKernel{K<:Kernel,T<:AbstractMatrix} <: MOKernel
    kernel::K
    B::T

    function CoregionMOKernel{K,T}(kernel::K, B::T) where {K,T}
        @check_args(CoregionMOKernel, isposdef(B), "B is pd") # it seems at least psd is needed?
        return new{K,T}(kernel, B)
    end
end

function CoregionMOKernel(kernel::Kernel, B::AbstractMatrix)
    return CoregionMOKernel{typeof(kernel),typeof(B)}(kernel, B)
end
```
In a second step one can consider if it is reasonable to add additional outer constructors such as the example given above. E.g., one could use `SymWoodbury` from [WoodburyMatrices](https://github.com/timholy/WoodburyMatrices.jl).

**Comment by 4aHxKzD** [comment]
Thank you! I have one question/"source of confusion", the elements of `B` are going to be parameters when we use this kernel in the context of GPs, so we are going to optimize them. If the kernel works with `B` directly, how do we ensure properties like `rank`, i.e., what stops the minimizer from returning a `B` that doesn't respect our requirements.

**Comment by devmotion** [comment]
You have to parameterize the matrix accordingly. It's similar to the case of positive parameters (see e.g. https://juliagaussianprocesses.github.io/AbstractGPs.jl/dev/examples/regression_1d/)  where you have to choose some parameterization (exp, logexp1, ...).

**Comment by 4aHxKzD** [comment]
something like this?
``` julia
function (ℓ::NegativeGPLoglikelihood)(params)
    kernel = ScaledKernel(
        transform(
            Matern52Kernel(),
            ScaleTransform(softplus(params[1]))
        ),
        softplus(params[2]),
    )
    A = params[3]
    v = params[4]
    B = A * transpose(A) + Diagonal(v)
    mokernel = CoregionMOKernel(kernel, B)
    f = GP(mokernel)
    fx = f(ℓ.x, 0.1)
    return -logpdf(fx, ℓ.y)
end
```

**Comment by devmotion** [comment]
Yes, something like this. And with an additional outer constructor it would be even shorter (since the constructor would compute `B` from `A` and `v`).

**Comment by 4aHxKzD** [comment]
Ok, i'm going to change and submit a PR.

should I use `isposdef` or something like `all(eigvals(B) .>= 0.0)`? Considering we only need PSD, and not PD.

**Comment by devmotion** [comment]
Since it's quite expensive, maybe we shouldn't check it (or at least make it possible to disable the checks, similar to Distributions.jl) :smile: In any case, we should rather use `eigmin(B) >= 0` than `all(eigvals(B) .>= 0.0)`.

An alternative would also be to restrict it to `PDMats.AbstractPDMat` - it also supports positive semi-definite (and Woodbury) matrices via [`PDMatsExtras.jl`](https://github.com/invenia/PDMatsExtras.jl).

**Comment by 4aHxKzD** [comment]
I think that this issue can be closed (#263 ).
